From de9f9062c0999145e10c7d81d516658a7e391360 Mon Sep 17 00:00:00 2001
From: Robert Beckett <bob.beckett@collabora.com>
Date: Fri, 11 Apr 2025 18:39:26 +0100
Subject: [PATCH 34/36] ply-renderer: add interface to allow head scraping

implementors should read back the pixel data from the head
in to a pixel buffer.

Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
---
 src/libply-splash-core/ply-renderer-plugin.h |  2 ++
 src/libply-splash-core/ply-renderer.c        | 14 ++++++++++++++
 src/libply-splash-core/ply-renderer.h        |  2 ++
 3 files changed, 18 insertions(+)

diff --git a/src/libply-splash-core/ply-renderer-plugin.h b/src/libply-splash-core/ply-renderer-plugin.h
index a3cbd4a3..36ee8696 100644
--- a/src/libply-splash-core/ply-renderer-plugin.h
+++ b/src/libply-splash-core/ply-renderer-plugin.h
@@ -77,6 +77,8 @@ typedef struct
         bool (*get_capslock_state)(ply_renderer_backend_t *backend);
         const char * (*get_keymap)(ply_renderer_backend_t *backend);
         void (*retain)(ply_renderer_backend_t *backend);
+        ply_pixel_buffer_t * (*scrape_buffer_from_head)(ply_renderer_backend_t *backend,
+                                                        ply_renderer_head_t    *head);
 } ply_renderer_plugin_interface_t;
 
 #endif /* PLY_RENDERER_PLUGIN_H */
diff --git a/src/libply-splash-core/ply-renderer.c b/src/libply-splash-core/ply-renderer.c
index 77625b0d..4f16052c 100644
--- a/src/libply-splash-core/ply-renderer.c
+++ b/src/libply-splash-core/ply-renderer.c
@@ -360,6 +360,20 @@ ply_renderer_get_buffer_for_head (ply_renderer_t      *renderer,
                                                                 head);
 }
 
+ply_pixel_buffer_t *
+ply_renderer_scrape_head (ply_renderer_t      *renderer,
+                          ply_renderer_head_t *head)
+{
+        assert (renderer != NULL);
+        assert (renderer->plugin_interface != NULL);
+        assert (head != NULL);
+
+        if (!renderer->plugin_interface->scrape_buffer_from_head)
+                return NULL;
+        return renderer->plugin_interface->scrape_buffer_from_head(renderer->backend,
+                                                                   head);
+}
+
 void
 ply_renderer_flush_head (ply_renderer_t      *renderer,
                          ply_renderer_head_t *head)
diff --git a/src/libply-splash-core/ply-renderer.h b/src/libply-splash-core/ply-renderer.h
index 11fa9b62..4267aa8c 100644
--- a/src/libply-splash-core/ply-renderer.h
+++ b/src/libply-splash-core/ply-renderer.h
@@ -65,6 +65,8 @@ const char *ply_renderer_get_device_name (ply_renderer_t *renderer);
 ply_list_t *ply_renderer_get_heads (ply_renderer_t *renderer);
 ply_pixel_buffer_t *ply_renderer_get_buffer_for_head (ply_renderer_t      *renderer,
                                                       ply_renderer_head_t *head);
+ply_pixel_buffer_t *ply_renderer_scrape_head (ply_renderer_t      *renderer,
+                                              ply_renderer_head_t *head);
 
 void ply_renderer_flush_head (ply_renderer_t      *renderer,
                               ply_renderer_head_t *head);
-- 
2.47.1

